File: erpnext/controllers/buying.py
# ERPNext - web based ERP (http://erpnext.com)
# Copyright (C) 2012 Web Notes Technologies Pvt Ltd
# 
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

from __future__ import unicode_literals
import webnotes
import webnotes.model

from webnotes.model.controller import DocListController
class BuyingController(DocListController):
	def validate_items(self, parentfield):
		for child in self.doclist.get({"parentfield": parentfield}):
			# get item controller. Also raises error if item not found
			itemdoc = webnotes.model.get("Item", child.item_code)[0]
			
			# check if purchase item
			if itemdoc.is_purchase_item != "Yes":
				webnotes.msgprint("""Item "%s" is not a Purchase Item""", raise_exception=True)

			# check if end of life has reached
			if itemdoc.end_of_life and getdate(itemdoc.end_of_life) <= \
					now_datetime().date():
				import stock
				webnotes.msgprint("""Item "%s" has reached its end of life""",
					raise_exception=stock.ItemEndOfLifeError)

			# check if warehouse is required
			# if itemdoc.is_stock_item == "Yes" and not child.warehouse:
			# 	webnotes.msgprint("""Warehouse is Mandatory for Item "%s", 
			# 		as it is a Stock Item""" % \
			# 		child.item_code, raise_exception=webnotes.MandatoryError)
			
	def validate_previous_doclist(self, parentfield, fieldname, item_fieldname):
		"""Validates current item row against previous item row"""
		doctypelist = webnotes.model.get_doctype(self.doc.doctype)
		prev_doctype = doctypelist.get_field(fieldname, parentfield=parentfield).options
		
		prev_doclist_cache = {}
		def get_prev_doclist(name):
			if not prev_doclist_cache.get(name):
				prev_doclist_cache[name] = webnotes.model.get(prev_doctype, name)
			return prev_doclist_cache.get(name)
		
		def errmsg(errfield, item, prev_item):
			label = doctypelist.get_label(errfield, parent=item.doctype)
			webnotes.msgprint("""%s of row # %d in the %s table
				cannot be different from the one in row # %d in the %s table 
				of %s: "%s" """ % (label, item.idx, item.doctype, prev_item.idx, 
				prev_item.doctype, prev_doctype, item.get(fieldname)),
				raise_exception=True)
		
		for item in self.doclist.get({"parentfield": parentfield}):
			if item.get(fieldname):
				prev_item = get_prev_doclist(item.get(fieldname)).getone({
					"name": item.get(item_fieldname)})
				
				if item.item_code != prev_item.item_code:
					# mismatch in item code
					errmsg("item_code", item, prev_item)
				if item.warehouse != prev_item.warehouse:
					# mismatch in warehouse
					errmsg("warehouse", item, prev_item)
				if item.uom != prev_item.uom:
					# mismatch in UOM
					errmsg("uom", item, prev_item)

File: public/js/buying.js
// ERPNext - web based ERP (http://erpnext.com)
// Copyright (C) 2012 Web Notes Technologies Pvt Ltd
// 
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

wn.require("app/js/transaction.js")
wn.provide("erpnext");

erpnext.Buying = erpnext.Transaction.extend({
	toggle_fields: function() {
		this.frm.toggle_display("contact_section", this.frm.doc.supplier);
	},
	
	// item_code: function(doc, cdt, cdn) {
	// 		var me = this;
	// 		var item = locals[cdt][cdn];
	// 		if(item.item_code) {
	// 			wn.call({
	// 				doc: this.frm.doc,
	// 				method: "get_item_details",
	// 				args: {
	// 					item_code: item.item_code,
	// 					warehouse: item.warehouse,
	// 				},
	// 				callback: function(r) {
	// 					// update item doc
	// 					$.extend(locals[cdt][cdn], r.message);
	// 					refresh_field(me.item_table_field);
	// 				}
	// 			})
	// 		}
	// 	},
	
	warehouse: function(doc, cdt, cdn) {
		var me = this;
		var item = locals[cdt][cdn];
		if(item.item_code && item.warehouse) {
			wn.call({
				method: "stock.get_projected_qty",
				args: {item_code: item.item_code, warehouse: item.warehouse},
				callback: function(r) {
					// TODO: how to refresh only a single grid cell
					item.projected_qty = r.message.projected_qty;
					refresh_field(me.item_table_field);
				}
			});
		}
	},
	
	uom: function(doc, cdt, cdn) {
		// QUESTION if there is an existing value, won't this overwrite them?
		var me = this;
		var item = locals[cdt][cdn];
		if(item.item_code && item.uom) {
			// update item doc with conversion factor, qty and rate
			wn.call({
				doc: this.frm.doc,
				method: "get_uom_details",
				args: {
					item_code: item.item_code,
					uom: item.uom,
					stock_qty: item.stock_qty,
					// qty: item.qty,
					exchange_rate: me.frm.doc.exchange_rate,
					doc_name: me.frm.doc.name,
				},
				callback: function(r) {
					$.extend(locals[cdt][cdn], r.message);
					refresh_field(me.item_table_field);

					// calculate
					// me.calc_amount();
				}
			});
		}
	},
	
	supplier: function() {
		var me = this;
		
		if(this.frm.doc.supplier) {
			wn.call({
				doc: this.frm.doc,
				method: "get_supplier_details",
				args: { supplier: me.frm.doc.supplier },
				callback: function(r) {
					me.frm.refresh();
				}
			});
		}
	},
	
	supplier_address: function() {
		if(this.frm.doc.supplier_address) {
			args = {"name": this.frm.doc.supplier_address};
			this.get_address(args);
		}
	},
	
	contact_person: function() {
		if(this.frm.doc.contact_person) {
			this.get_contact();
		}
	},
	
	is_stopped: function() {
		if(this.frm.doc.docstatus != 1) {
			var msg = repl("%(doctype)s must be submitted to stop/resume it",
				this.frm.doc);
			msgprint(msg);
			throw msg;
		}
	},
	
	set_schedule_date: function(item_table_field) {
		// if children, then set schedule date based on lead time days
		if(wn.model.has_children(this.frm.doc.doctype, this.frm.doc.name,
				item_table_field)) {
			wn.call({
				doc: this.frm.doc,
				method: "set_schedule_date",
				args: { item_table_field: item_table_field },
				callback: function(r) {
					refresh_field(item_table_field);
				}
			})	
		}
	},
	
	setup_get_query: function() {
		this._super();
		
		var me = this;
		var item_grid = this.frm.fields_dict[this.item_table_field].grid;
		
		// item_code
		item_grid.get_field("item_code").get_query = function() {
			if(me.frm.doc.is_subcontracted == "Yes") {
				var condition = "ifnull(is_subcontracted_item, 'No')='Yes'"
			} else {
				var condition = "ifnull(is_purchase_item, 'No')='Yes'"
			}
			var query = "select name, description from `tabItem` \
				where " + condition + " \
				and (ifnull(end_of_life, '')='' or end_of_life='0000-00-00' \
				or end_of_life > now()) and docstatus!=2 \
				and %(key)s like \"%s\" limit 50";
			
			return query;
		};
		
		// project
		// TODO: should be filtered by customer or supplier!
		if(item_grid.get_field("project_name")) {
			item_grid.get_field("project_name").get_query = function() {
				return "select name from `tabProject` \
					where status not in ('Completed', 'Cancelled') \
					and %(key)s like \"%s\" order by name asc limit 50";
			};
		}
		
		// supplier address
		if(this.frm.fields_dict.supplier_address) {
			this.frm.fields_dict.supplier_address.get_query = function() {
				return repl("select name, address_line1, city from `tabAddress` \
					where supplier = \"%(supplier)s\" and docstatus < 2 \
					and %(key)s like \"%s\" order by name asc limit 50", {
						supplier: me.frm.doc.supplier,
					});
			}
			this.frm.fields_dict.supplier_address.on_new = function(name) {
				me.set_dialog_defaults("Address", name);
			}
		}
		
		
		// contact person
		if(this.frm.fields_dict.contact_person) {
			this.frm.fields_dict.contact_person.get_query = function() {
				return repl("select name, concat(first_name, ' ', ifnull(last_name, '')) \
					as fullname, department, designation from tabContact \
					where supplier = \"%(supplier)s\" and docstatus < 2 \
					and %(key)s like \"%s\" order by name asc limit 50", {
						supplier: me.frm.doc.supplier,
					});
			}
			this.frm.fields_dict.contact_person.on_new = function(name) {
				me.set_dialog_defaults("Contact", name);
			}
		}
	},
	
	set_dialog_defaults: function(doctype, name) {
		var doc = locals[doctype][name];
		doc.supplier = me.frm.doc.supplier;
		doc.supplier_name = me.frm.doc.supplier_name;
	},
});